### ISSUE TITLE
Shape mismatch of target array when using categorical_crossentropy and stateful RNNs

### ISSUE THREAD

**Issue by chenguangshen** [fchollet/keras#2571]
Hi guys,

I'm training an RNN model on a time-series classification task. The training set I have is of the shape (200000, 50, 3), i.e. each input series have `input_length = 50` and `input_dim = 3`, and I have 200000 such series. The task is to classify each input series with a label 1-4 (4 classes). Each input series will be tagged with one and only one label. After one-hot transformation the target array (ground truth labels) is of shape (200000, 4), which contains only binary integers.

Since I'm setting `stateful=true` for `SimpleRNN`, I have to give the full input shape, with a batch size of 120. My model construction is shown below:
```python
model = Sequential()
model.add(SimpleRNN(n_hidden, batch_input_shape=(120, 50, 3), stateful=True, return_sequences=True))
model.add(Dropout(dropout))
model.add(SimpleRNN(n_hidden, stateful=True, return_sequences=True))
model.add(Dropout(dropout))
model.add(TimeDistributed(Dense(4)))
model.add(Activation('softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam')
```

I tried to fit the above model with samples (200000, 50, 3) and target array (200000, 4) but got the error:
```
Exception: A target array with shape (200000, 4) was passed for an output of shape (120, 50, 4) while using as loss `categorical_crossentropy`. This loss expects targets to have the same shape as the output.
```

I also tried reshaping the target array to (200000, 1, 4), or repeating it to be of shape (200000, 50, 4) but the error would become:
```
Exception: You are passing a target array of shape (200000, 1, 4) while using as loss `categorical_crossentropy`. `categorical_crossentropy` expects targets to be binary matrices (1s and 0s) of shape (samples, classes).
```
Note that other loss functions, e.g. hinge or mse, work well with the (200000, 1, 4) target array.

I get that `stateful=true` requires full input shape for RNNs. What I don't understand is that why it is trying to fit the target array with the same `input_length = 50` as the training samples. Each batch (120, 50, 3) should be fitted with a (120, 4) target array instead of a (120, 50, 4) target array, because each input series only have one corresponding label.

Any suggestions or pointers will be much appreciated. Thanks!

**Comment by joelthchao** [comment]
Please refer to [stateful LSTM example](https://github.com/fchollet/keras/blob/master/examples/stateful_lstm.py), the last `SimpleRNN` should be set `return_sequences=False` for classification task and `TimeDistributed` is not needed.
```python
model = Sequential()
model.add(SimpleRNN(n_hidden, batch_input_shape=(120, 50, 3), stateful=True, return_sequences=True))
model.add(Dropout(dropout))
model.add(SimpleRNN(n_hidden, stateful=True, return_sequences=False))
model.add(Dropout(dropout))
model.add(Dense(4))
model.add(Activation('softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam')

# x shape: (nb_samples, 50, 3)
# y shape: (nb_samples, 4)

model.fit(x, y, batch_size=120)
```

**Comment by chenguangshen** [comment]
@joelthchao thanks so much for your help!
